x-pack/plugins/apm/public/components/alerting/utils/helper.ts

```diff
@@ -6,10 +6,12 @@
  */
 
 import { TIME_UNITS } from '@kbn/triggers-actions-ui-plugin/public';
+import type { RuleTypeMetaData } from '@kbn/alerting-plugin/common';
+
 import moment from 'moment';
 
-export interface AlertMetadata {
-  environment: string;
+export interface AlertMetadata extends RuleTypeMetaData {
+  environment?: string;
```

Comment:
Why did you make `environment` optional here?

Reply:
This is as-is behaviour. It was previously masked by the fact that you casted `metadata` into type `AlertMetadata` which pretended that `environment` exists despite it being possibly undefined in the code:

https://github.com/elastic/kibana/pull/175777/files/eadf7094eb90366fa539d8586374898713e58130#diff-d3c61c7985f7b6630da96a8ca74b6cfee9069f8e9d00bbf176e43fd95d0da4f9R38-R39

I've now made this explicit. 

If environment is a required prop and should always be defined then you need to change the implementation above and provide a fallback for the scenario when `query.environment` is `undefined`.

Reply:
I looked at the code, and `environment` has been set everywhere where an alert can be created.

>If environment should always be defined then we need to change the implementation above and provide a fallback for the scenario when query.environment is undefined.

The piece of code you linked above handles the fallback, right? `environment` will be undefined for the settings page for example, where it is not required, in this case, it falls back to `ALL`. For all the other pages environment will always be defined.

I think we should not make `environment` optional here.

Reply:
> The piece of code you linked above handles the fallback, right? `environment` will be undefined for the settings page for example, where it is not required, in this case, it falls back to `ALL`. For all the other pages environment will always be defined.

That's not what the code does: 

```ts
const environment = 'environment' in query ? query.environment : ENVIRONMENT_ALL.value;
```

This snippet only checks whether a key with the name 'environment' exists. It does not check whether its value is truthy which is why `query.environment` can still be undefined. 

> I think we should not make `environment` optional here.

Ok I will fix the logic accordingly

Reply:
Ah nice, the TS compiler is currently not aware of that though so the `query` type might need to be adjusted to reflect this logic.

Reply:
That's actually because we use `/*` here https://github.com/elastic/kibana/blob/main/x-pack/plugins/apm/public/components/alerting/ui_components/alerting_flyout/index.tsx#L30

This makes the environment optional since we have a few pages where the environment is not necessary, such as the Settings page. In this case, it'll fallback to the environment `ALL`

Reply:
> That's actually because we use `/*` here https://github.com/elastic/kibana/blob/main/x-pack/plugins/apm/public/components/alerting/ui_components/alerting_flyout/index.tsx#L30
> 
> This makes the environment optional since we have a few pages where the environment is not necessary, such as the Settings page. In this case, it'll fallback to the environment `ALL`

I don't understand. Is `environment` optional in this component or not?

Reply:
>I don't understand. Is environment optional in this component or not?

No, there's no `environment` on the Settings page at all, so:

```
'environment' in query
```

will be false.

Reply:
Got you! :) Cool, I'll mark this as resolved then.